File: TJ_FX_ROBOT_CONTRL_SDK-master/DEMO_C++/showcase_validate_get_dcss_data_from_user_inputCMD.cpp
#include "MarvinSDK.h"
#include "stdio.h"
#include "stdlib.h"
#include "unistd.h"
#include <iostream>
#include <cstdlib>
#include <cstdio>

#ifdef _WIN32
#include <windows.h>
#define SLEEP(ms) Sleep(ms)
#else
#include <unistd.h>
#define SLEEP(ms) usleep((ms) * 1000)
#endif

// '''#################################################################
// 该 DEMO 用于【左臂：设置笛卡尔阻抗/力控相关参数 + 切换扭矩模式 + 读回校验】的示例。

// '''#################################################################
int main()
{
    // 初始化订阅数据的结构体
    DCSS dcss;

    // 查验连接是否成功
    bool init = OnLinkTo(192, 168, 1, 190);
    if (!init)
    {
        std::cerr << "failed to connect to the robot, port is occupied" << std::endl;
        return -1;
    }

    SLEEP(200);
    // 检查伺服和手臂是否有错，有错误清错
    // 订阅最新数据获取机械臂的错误和状态，有错误清错
    OnGetBuf(&dcss);
    int arm_error_a = dcss.m_State[0].m_ERRCode;
    int arm_error_b = dcss.m_State[1].m_ERRCode;
    int arm_state_a = dcss.m_State[0].m_CurState;
    int arm_state_b = dcss.m_State[1].m_CurState;
    if (arm_error_a != 0 || arm_state_a == 100)
    {
        std::cout << "arm A: exits error, clear error\n"
                  << std::endl;
        SLEEP(20);
        OnClearSet();
        OnClearErr_A();
        OnSetSend();
        SLEEP(20);
    }
    if (arm_error_b != 0 || arm_state_b == 100)
    {
        std::cout << "arm B: exits error, clear error\n"
                  << std::endl;
        SLEEP(20);
        OnClearSet();
        OnClearErr_B();
        OnSetSend();
        SLEEP(20);
    }

    // 获取伺服错误，有错误清错
    long ErrCode_A[7] = {};
    long ErrCode_B[7] = {};
    OnGetServoErr_A(ErrCode_A);
    OnGetServoErr_B(ErrCode_B);
    bool allZero_a = true;
    bool allZero_b = true;
    for (int i = 0; i < 7; ++i)
    {
        if (ErrCode_A[i] != 0)
        {
            allZero_a = false;
            break;
        }
    }
    for (int i = 0; i < 7; ++i)
    {
        if (ErrCode_B[i] != 0)
        {
            allZero_b = false;
            break;
        }
    }
    if (!allZero_a)
    {
        std::cout << "arm A: srvo error exists, clear error\n"
                  << std::endl;
        SLEEP(20);
        OnClearSet();
        OnClearErr_A();
        OnSetSend();
        SLEEP(20);
    }
    if (!allZero_b)
    {
        std::cout << "arm B: srvo error exists, clear error\n"
                  << std::endl;
        SLEEP(20);
        OnClearSet();
        OnClearErr_B();
        OnSetSend();
        SLEEP(20);
    }

    // 通过确认freame数据的刷新，确认UDP数据通道连接成功（防火墙等可能不能正常收到数据）
    int motion_tag = 0;
    int frame_update = 0;

    for (int i = 0; i < 5; i++)
    {
        OnGetBuf(&dcss);
        std::cout << "connect frames:" << dcss.m_Out[0].m_OutFrameSerial << std::endl;

        if (dcss.m_Out[0].m_OutFrameSerial != 0 &&
            frame_update != dcss.m_Out[0].m_OutFrameSerial)
        {
            motion_tag++;
            frame_update = dcss.m_Out[0].m_OutFrameSerial;
        }
        SLEEP(1);
    }
    if (motion_tag > 0)
    {
        std::cout << "success:robot connected\n"
                  << std::endl;
    }
    else
    {
        std::cerr << "failed:robot connection failed\n"
                  << std::endl;
        OnRelease();
        return -1;
    }

    // 控制日志开
    OnLogOn();
    OnLocalLogOn();

    int i = 5;
    OnSetUserSpcfData_A(DCSS_CMD_ARM0_GET_DATA_6FT);
    i = 5;

    while (i-- > 0)
    {

        OnGetBuf(&dcss);
        printf("DCSS_CMD_ARM0_GET_DATA_6FT [%lf %lf %lf %lf %lf %lf]\n", dcss.m_Out[0].m_EST_Joint_Firc_Dot[0],
               dcss.m_Out[0].m_EST_Joint_Firc_Dot[1],
               dcss.m_Out[0].m_EST_Joint_Firc_Dot[2],
               dcss.m_Out[0].m_EST_Joint_Firc_Dot[3],
               dcss.m_Out[0].m_EST_Joint_Firc_Dot[4],
               dcss.m_Out[0].m_EST_Joint_Firc_Dot[5]);
        printf("current specify Data = %f \n", dcss.m_Out[0].m_EST_Joint_Firc[0]);
        SLEEP(500);
    }

    OnSetUserSpcfData_A(DCSS_CMD_ARM0_GET_DATA_GRAVITY);
    i = 5;
    while (i-- > 0)
    {
        OnGetBuf(&dcss);
        printf("DCSS_CMD_ARM0_GET_DATA_GRAVITY [%lf %lf %lf %lf %lf %lf %lf]\n", dcss.m_Out[0].m_EST_Joint_Firc_Dot[0],
               dcss.m_Out[0].m_EST_Joint_Firc_Dot[1],
               dcss.m_Out[0].m_EST_Joint_Firc_Dot[2],
               dcss.m_Out[0].m_EST_Joint_Firc_Dot[3],
               dcss.m_Out[0].m_EST_Joint_Firc_Dot[4],
               dcss.m_Out[0].m_EST_Joint_Firc_Dot[5],
               dcss.m_Out[0].m_EST_Joint_Firc_Dot[6]);
        printf("current specify Data = %f \n", dcss.m_Out[0].m_EST_Joint_Firc[0]);
        SLEEP(500);
    }

    OnSetUserSpcfData(DCSS_CMD_GET_DATA_GYRO_ACC);
    i = 5;
    while (i-- > 0)
    {
        OnGetBuf(&dcss);
        printf("DCSS_CMD_GET_DATA_GYRO_ACC [%lf %lf %lf]\n", dcss.m_Out[0].m_EST_Cart_FN[0],
               dcss.m_Out[0].m_EST_Cart_FN[1],
               dcss.m_Out[0].m_EST_Cart_FN[2]);
        printf("current specify Data = %f \n", dcss.m_Out[0].m_EST_Joint_Firc[1]);
        printf("current specify Data = %f \n", dcss.m_Out[1].m_EST_Joint_Firc[1]);
        SLEEP(500);
    }

    OnSetUserSpcfData(DCSS_CMD_GET_DATA_GYRO_ANGLE);
    i = 5;
    while (i-- > 0)
    {
        OnGetBuf(&dcss);
        printf("DCSS_CMD_GET_DATA_GYRO_ANGLE [%lf %lf %lf]\n", dcss.m_Out[0].m_EST_Cart_FN[0],
               dcss.m_Out[0].m_EST_Cart_FN[1],
               dcss.m_Out[0].m_EST_Cart_FN[2]);
        printf("current specify Data = %f \n", dcss.m_Out[0].m_EST_Joint_Firc[1]);
        printf("current specify Data = %f \n", dcss.m_Out[1].m_EST_Joint_Firc[1]);
        SLEEP(500);
    }

    OnSetUserSpcfData(DCSS_CMD_GET_DATA_GYRO_OMEG);
    i = 5;
    while (i-- > 0)
    {
        OnGetBuf(&dcss);
        printf("DCSS_CMD_GET_DATA_GYRO_OMEG [%lf %lf %lf]\n", dcss.m_Out[0].m_EST_Cart_FN[0],
               dcss.m_Out[0].m_EST_Cart_FN[1],
               dcss.m_Out[0].m_EST_Cart_FN[2]);
        printf("current specify Data = %f \n", dcss.m_Out[0].m_EST_Joint_Firc[1]);
        printf("current specify Data = %f \n", dcss.m_Out[1].m_EST_Joint_Firc[1]);
        SLEEP(500);
    }

    OnRelease();
    return 1;
}

File: TJ_FX_ROBOT_CONTRL_SDK-master/DEMO_PYTHON/showcase_validate_get_dcss_data_from_user_inputCMD.py
import sys
import os
current_dir = os.path.dirname(os.path.abspath(__file__))
parent_dir = os.path.dirname(current_dir)
sys.path.insert(0, parent_dir)
current_file_path = os.path.abspath(__file__)
current_path = os.path.dirname(current_file_path)
from SDK_PYTHON.fx_robot import Marvin_Robot, DCSS
import time
import logging

'''#################################################################
该 DEMO 演示 通过 OnSetUserSpcfData 接口切换用户自定义数据通道，
将不同的机器人内部数据（六维力、重力、陀螺仪等）映射到 DCSS 的固定字段中，
从而在不改变结构体定义的前提下读取多样的内部状态。

使用逻辑
     连接机器人 + 清错 + 验证数据通道
     切换数据类别 → 订阅读取 → 打印对应的 DCSS 字段
     支持 arm='A' / 'B' / 'AB'(双臂)
     释放机器人
'''#################################################################

# 配置日志系统
logging.basicConfig(format='%(message)s')
logger = logging.getLogger('debug_printer')
logger.setLevel(logging.INFO)   # 一键关闭所有调试打印
logger.setLevel(logging.DEBUG)  # 默认开启DEBUG级

# ── 用户自定义数据类别常量 ──
DCSS_CMD_ARM0_GET_DATA_6FT     = 116   # A臂: 六维力数据
DCSS_CMD_ARM0_GET_DATA_GRAVITY = 117   # A臂: 重力数据
DCSS_CMD_GET_DATA_GYRO_ACC     = 300   # 双臂: 陀螺仪加速度
DCSS_CMD_GET_DATA_GYRO_ANGLE   = 301   # 双臂: 陀螺仪角度
DCSS_CMD_GET_DATA_GYRO_OMEG    = 302   # 双臂: 陀螺仪角速度

'''初始化订阅数据的结构体'''
dcss = DCSS()

'''初始化机器人接口'''
robot = Marvin_Robot()

'''查验连接是否成功'''
init = robot.connect('192.168.1.190')
if init == 0:
    logger.error('failed to connect to the robot, port is occupied')
    exit(0)

time.sleep(0.2)

'''检查伺服和手臂是否有错，有错误清错'''
robot.check_error_and_clear(dcss)

'''通过确认frame数据的刷新，确认UDP数据通道连接成功'''
motion_tag = 0
frame_update = None
for i in range(5):
    sub_data = robot.subscribe(dcss)
    print(f"connect frames :{sub_data['outputs'][0]['frame_serial']}")
    if sub_data['outputs'][0]['frame_serial'] != 0 and frame_update != sub_data['outputs'][0]['frame_serial']:
        motion_tag += 1
        frame_update = sub_data['outputs'][0]['frame_serial']
    time.sleep(0.001)

if motion_tag > 0:
    logger.info('success:robot connected')
else:
    logger.error('failed:robot connection failed')
    robot.release_robot()
    exit(0)

'''控制日志开'''
robot.log_switch('1')
robot.local_log_switch('1')

# ═══════════════════════════════════════════════════════════════════
# 1. DCSS_CMD_ARM0_GET_DATA_6FT — A臂读取六维力数据
# ═══════════════════════════════════════════════════════════════════
robot.set_user_specified_data('A', DCSS_CMD_ARM0_GET_DATA_6FT)
i = 5
while i > 0:
    i -= 1
    sub_data = robot.subscribe(dcss)
    out0 = sub_data['outputs'][0]
    print(f"DCSS_CMD_ARM0_GET_DATA_6FT {out0['est_joint_firc_dot'][:6]}")
    print(f"current specify Data = {out0['est_joint_firc'][0]}")
    time.sleep(0.5)

# ═══════════════════════════════════════════════════════════════════
# 2. DCSS_CMD_ARM0_GET_DATA_GRAVITY — A臂读取重力数据
# ═══════════════════════════════════════════════════════════════════
robot.set_user_specified_data('A', DCSS_CMD_ARM0_GET_DATA_GRAVITY)
i = 5
while i > 0:
    i -= 1
    sub_data = robot.subscribe(dcss)
    out0 = sub_data['outputs'][0]
    print(f"DCSS_CMD_ARM0_GET_DATA_GRAVITY {out0['est_joint_firc_dot'][:7]}")
    print(f"current specify Data = {out0['est_joint_firc'][0]}")
    time.sleep(0.5)

# ═══════════════════════════════════════════════════════════════════
# 3. DCSS_CMD_GET_DATA_GYRO_ACC — 双臂读取陀螺仪加速度
# ═══════════════════════════════════════════════════════════════════
robot.set_user_specified_data('C', DCSS_CMD_GET_DATA_GYRO_ACC)
i = 5
while i > 0:
    i -= 1
    sub_data = robot.subscribe(dcss)
    out0 = sub_data['outputs'][0]
    out1 = sub_data['outputs'][1]
    print(f"DCSS_CMD_GET_DATA_GYRO_ACC {out0['est_cart_fn'][:3]}")
    print(f"current specify Data A = {out0['est_joint_firc'][1]}")
    print(f"current specify Data B = {out1['est_joint_firc'][1]}")
    time.sleep(0.5)

# ═══════════════════════════════════════════════════════════════════
# 4. DCSS_CMD_GET_DATA_GYRO_ANGLE — 双臂读取陀螺仪角度
# ═══════════════════════════════════════════════════════════════════
robot.set_user_specified_data('AB', DCSS_CMD_GET_DATA_GYRO_ANGLE)
i = 5
while i > 0:
    i -= 1
    sub_data = robot.subscribe(dcss)
    out0 = sub_data['outputs'][0]
    out1 = sub_data['outputs'][1]
    print(f"DCSS_CMD_GET_DATA_GYRO_ANGLE {out0['est_cart_fn'][:3]}")
    print(f"current specify Data A = {out0['est_joint_firc'][1]}")
    print(f"current specify Data B = {out1['est_joint_firc'][1]}")
    time.sleep(0.5)

# ═══════════════════════════════════════════════════════════════════
# 5. DCSS_CMD_GET_DATA_GYRO_OMEG — 双臂读取陀螺仪角速度
# ═══════════════════════════════════════════════════════════════════
robot.set_user_specified_data('AB', DCSS_CMD_GET_DATA_GYRO_OMEG)
i = 5
while i > 0:
    i -= 1
    sub_data = robot.subscribe(dcss)
    out0 = sub_data['outputs'][0]
    out1 = sub_data['outputs'][1]
    print(f"DCSS_CMD_GET_DATA_GYRO_OMEG {out0['est_cart_fn'][:3]}")
    print(f"current specify Data A = {out0['est_joint_firc'][1]}")
    print(f"current specify Data B = {out1['est_joint_firc'][1]}")
    time.sleep(0.5)

'''释放机器人内存'''
robot.release_robot()
